text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,581 @@
+    public static UnionType fromFoo(Foo value) {
+      return new UnionType.FooUnionType(value);
+    }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) {
+      return new UnionType.MapOfBarUnionType(value);
+    }
+    public static UnionType fromListOfBaz(Baz... value) {
+      return fromListOfBaz(Arrays.asList(value));
+    }
+    public static UnionType fromListOfBaz(List<Baz> value) {
+      return new UnionType.ListOfBaz(value);
+    }
+
+    public @Nullable Foo asFoo() { return null; }
+    public @Nullable Map<String, Bar> asMapOfBar() { return null; }
+    public @Nullable List<Baz> asListOfBaz() { return null; }
+
+    // Private constructor ensures all initialization happens through the factories
+    private UnionType(/* ... */) { /* ... */ }
+
+    private static final class FooUnionType extends UnionType { /*...*/ }
+    private static final class MapOfBarUnionType extends UnionType { /*...*/ }
+    private static final class ListOfBazUnionType extends UnionType { /*...*/ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a new argument to the _jsii
+kernel_'s `load` API if it's code was generated with `--union-like-types` enabled. This way, the kernel will be bale to reference the declared _type
+union_'s information when annotating instances that are passed from **Javascript** to the host application, removing all ambiguity regarding which
+type is serialized.
+
+# Drawbacks
+
+This approach may allow developers in statically typed languages to cause _undefined behavior_ to occur if they are not provided with a way to
+validate whether a _union-like_ instancs is of a candidate type or not. Without such an API, developers may use an invalid `as<TypeName>` accessor and
+retrive a proxy that may behave unexpectedly. This behavior is however identical to what happens in **TypeScript** when developers use `as any` to
+evade type checking.
+
+# Rationale and Alternatives
+
+Improving support for _type unions_ appears to be the best way going forward, as complete removal of type unions has proven to be challenging from an
+API design pespective: it tends to make the experience significantly worse for dynamic languages (as boilerplate-heavy patterns must be leveraged as a
+replacement for _type unions_), and would reduce the flexibility in language-specific code generation for _union-like_ types (as those would be bound
+to the **TypeScript** API). Full support of _type unions_ as a type kind in the _jsii type model_ achieves better outcomes:
+
+- it ensures consistency around _union-like_ types
+- it does not degrade the developer experience in languages with _native type_ union support
+- it requires significantly less effort from library authors
+
+# Adoption Strategy
+
+Leveraging aliased _type unions_ in all existing CDK APIs would be a breaking change for all statically typed bindings. As such, the initial rollout
```

Comment:
Simpler proposed rollout strategy:

* Rewrite all current code to use **named unions** (most of them are generated so shouldn't be too hard).
* Add a pacmak flag to generate v1 bindings treating the unions as "inline types", and v2 bindings generating the union classes.

Reply:
This was the proposal before moving to that. The inconvenient of the approach you are proposing is that it does not allow fixing the **currently broken** things by generating better type signatures in Java/C#. Those methods could be broken right now since they cannot currently be used from those languages.

Reply:
Can you give examples of things that are today outright *broken* though? The only change that's acceptable is things that simply *cannot be used* today (*can be used but with ugly syntax* is not good enough since presumably we'd be changing the semantics).